File: Programmers/KaKao/2020_인턴십_키패드 누르기.java
class Solution {
    public String solution(int[] numbers, String hand) {
        StringBuilder answer = new StringBuilder();
        int[][] keypad = new int[12][12];
        int tempLeft = 10;
        int tempRight = 11;
        keypad[0] = new int[]{0, 4, 3, 4, 3, 2, 3, 2, 1, 2, 1, 1};
        keypad[1] = new int[]{4, 0, 1, 2, 1, 2, 3, 2, 3, 4, 3, 5};
        keypad[2] = new int[]{3, 1, 0, 1, 2, 1, 2, 3, 2, 3, 4, 4};
        keypad[3] = new int[]{4, 2, 1, 0, 3, 2, 1, 4, 3, 2, 5, 3};
        keypad[4] = new int[]{3, 1, 2, 3, 0, 1, 2, 1, 2, 3, 2, 4};
        keypad[5] = new int[]{2, 2, 1, 2, 1, 0, 1, 2, 1, 2, 3, 3};
        keypad[6] = new int[]{3, 3, 2, 1, 2, 1, 0, 3, 2, 1, 4, 2};
        keypad[7] = new int[]{2, 2, 3, 4, 1, 2, 3, 0, 1, 2, 1, 3};
        keypad[8] = new int[]{1, 3, 2, 3, 2, 1, 2, 1, 0, 1, 2, 2};
        keypad[9] = new int[]{2, 4, 3, 2, 3, 2, 1, 2, 1, 0, 3, 1};
        keypad[10] = new int[]{1, 3, 4, 5, 2, 3, 4, 1, 2, 3, 0, 2};
        keypad[11] = new int[]{1, 5, 4, 3, 4, 3, 2, 3, 2, 1, 2, 0};

        for(int number : numbers) {
            if(number == 1 || number == 4 || number == 7) {
                answer.append("L");
                tempLeft = number;
            }
            else if(number == 3 || number == 6 || number == 9) {
                answer.append("R");
                tempRight = number;
            }
            else {
                if(keypad[number][tempLeft] < keypad[number][tempRight]) {
                    answer.append("L");
                    tempLeft = number;
                }
                else if(keypad[number][tempLeft] > keypad[number][tempRight]) {
                    answer.append("R");
                    tempRight = number;
                }
                else if(hand.equals("left")) {
                    answer.append("L");
                    tempLeft = number;
                }
                else if(hand.equals("right")) {
                    answer.append("R");
                    tempRight = number;
                }
            }
        }


        return answer.toString();
    }
}

File: Programmers/KaKao/2020_인턴십_키패드 누르기.kt
import kotlin.text.StringBuilder

class Solution {
    val dist_2 = arrayOf(3, 1, 0, 1, 2, 1, 2, 3, 2, 3, 4, 4)
    val dist_5 = arrayOf(2, 2, 1, 2, 1, 0, 1, 2, 1, 2, 3, 3)
    val dist_8 = arrayOf(1, 3, 2, 3, 2, 1, 2, 1, 0, 1, 2, 2)
    val dist_0 = arrayOf(0, 4, 3, 4, 3, 2, 3, 2, 1, 2, 1, 1)
    var HAND = ""

    fun solution(numbers: IntArray, hand: String): String {

        var leftTemp = 10
        var rightTemp = 11
        HAND = hand
        var answer = StringBuilder()

        for(number in numbers) {
            when(number) {
                1, 4, 7 -> {
                    answer.append("L")
                    leftTemp = number
                }
                3, 6, 9 ->  {
                    answer.append("R")
                    rightTemp = number
                }
                else -> {
                    val near = checkDist(number, leftTemp, rightTemp)
                    answer.append(near)
                    if(near == "L") { leftTemp = number }
                    else { rightTemp = number }
                }
            }
        }

        return answer.toString()
    }

    fun checkDist(target: Int, leftTemp: Int, rightTemp: Int): String {
        var leftDist = 0
        var rightDist = 0

        when(target) {
            2 -> {
                leftDist = dist_2[leftTemp]
                rightDist = dist_2[rightTemp]
            }
            5 -> {
                leftDist = dist_5[leftTemp]
                rightDist = dist_5[rightTemp]
            }
            8 -> {
                leftDist = dist_8[leftTemp]
                rightDist = dist_8[rightTemp]
            }
            0 -> {
                leftDist = dist_0[leftTemp]
                rightDist = dist_0[rightTemp]
            }
        }

        if(leftDist == rightDist) {
            if(HAND == "left") { return "L" }
            else { return "R" }
        }
        else if (leftDist > rightDist) {
            return "R"
        }
        else {
            return "L"
        }
    }
}
